### PR TITLE
Ensure that it is not possible to explicitly free stack memory

### PR DESCRIPTION
**Pull request by oli-obk**
cc @RalfJung 

You had to write a workaround around my hack. Both are gone now :)

### CODE REVIEW

File: src/memory.rs [hunk 1]
```diff
@@ -34,19 +34,25 @@ pub struct Allocation {
...
-    /// may be modified, but never deallocated
-    Mutable,
-    /// may neither be modified nor deallocated
-    Immutable,
+pub enum Kind {
+    /// Error if deallocated any other way than `rust_deallocate`
+    Rust,
+    /// Error if deallocated any other way than `free`
+    C,
+    /// Error if deallocated via `rust_deallocate`
```
Comment (RalfJung) [inline src/memory.rs]: Also error if deallocated via `free`, I assume?

File: src/memory.rs [hunk 2]
```diff
@@ -627,26 +633,31 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
     }
 
     /// mark an allocation pointed to by a static as static and initialized
-    pub fn mark_inner_allocation(&mut self, alloc: AllocId, mutable: bool) -> EvalResult<'tcx> {
+    pub fn mark_inner_allocation(&mut self, alloc: AllocId, make_immutable: bool) -> EvalResult<'tcx> {
```
Comment (RalfJung) [inline src/memory.rs]: Such negation of what booleans mean makes me nervous. Wouldn't it be better to use a dedicated enum here?

File: tests/compile-fail/double_free.rs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+fn main() {
+    let x = Box::new(42);
+    {
+        let bad_box: Box<i32> = unsafe { std::ptr::read(&x) };
+        drop(bad_box);
+    }
+    drop(x); //~ ERROR dangling pointer was dereferenced
+}
```
Comment (RalfJung) [inline tests/compile-fail/double_free.rs]: This is also tested by deallocate-twice.rs.

**Comment by RalfJung** [comment]
Yay. :D (Also -- dang it, yet another PR to rebase [mir-validate](https://github.com/RalfJung/miri/tree/mir-validate) around. Why do all these PRs keep touching memory?^^)

However, I'm still not sure I understand why `deallocate_local` gets called for the same static multiple times.

File: src/memory.rs (hunk 2 above)
Comment (eddyb) [inline src/memory.rs]: We have such an enum in rustc already, in `rustc::hir` I believe.

File: tests/compile-fail/double_free.rs (hunk 3 above)
Comment (oli-obk) [inline tests/compile-fail/double_free.rs]: ah great! I'll drop the test then

**Comment by oli-obk** [comment]
> However, I'm still not sure I understand why deallocate_local gets called for the same static multiple times.

It doesn't. The issue is that the local can be referenced by the return value. So when we freeze the return value, we also freeze all the values that the return value has relocations into. Then afterwards, we free all locals that weren't frozen.

File: src/memory.rs (hunk 2 above)
Comment (oli-obk) [inline src/memory.rs]: and one in `syntax::ast` and one in `rustc::mir`... I think I need to make a PR to rustc...

**Comment by RalfJung** [comment]
Ah, makes sense. :)  Explain that in the code?

**Comment by RalfJung** [comment]
Other than that, 👍

File: src/memory.rs [hunk 4]
```diff
@@ -206,57 +218,54 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
...
-            if alloc.static_kind != StaticKind::NotStatic {
-                return Err(EvalError::DeallocatedStaticMemory);
-            }
-            if let Some((size, align)) = size_and_align {
-                if size != alloc.bytes.len() as u64 || align != alloc.align {
-                    return Err(EvalError::IncorrectAllocationInformation);
-                }
+        let alloc = match self.alloc_map.remove(&ptr.alloc_id) {
+            Some(alloc) => alloc,
+            None => return Err(EvalError::DoubleFree),
```
Comment (RalfJung) [inline src/memory.rs]: Actually I just noticed this line is unreachable -- we already check above `self.get(ptr.alloc_id).is_err()`. (There a similar redundant conditional in `reallocate`.)

Comment (oli-obk) [inline src/memory.rs]: We should remove that check. The double free error is much clearer imo

**Pull request merged**